update MAPE

raybellwaves opened the pull request:
# Description

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

Closes #160 

## Type of change

Please delete options that are not relevant.

-   [x]  refactoring

# How Has This Been Tested?

Moved the mape test out of the numpy custom functions to the scikit-learn tests.

raybellwaves commented:
believe the python 3.6 issues in the CI are outside this PR. the doc yaml is set at 3.6 which is probably why it's also failing.

aaronspring commented on xskillscore/core/np_deterministic.py:
```diff
@@ -652,7 +652,7 @@ def _mape(a, b, weights, axis, skipna):
         \\mathrm{MAPE} = \\frac{1}{n} \\sum_{i=1}^{n}
                          \\frac{\\vert a_{i} - b_{i} \\vert}
-                               {\\vert a_{i} \\vert}
+                               {max(\epsilon, \\vert a_{i} \\vert)}
```
can you also add this epsilon to other deterministic metrics? or should this be another PR?

aaronspring reviewed (approved):
code LGTM, somehow tests are failing

maybe you could also add a test for the epsilon case

raybellwaves commented:
> can you also add this epsilon to other deterministic metrics? or should this be another PR?

I vote leave for another PR. Could create an issue and list which metrics can be updated.

> tests are failing

Yes deffo something weird with numba and py3.6
```
/usr/share/miniconda/envs/xskillscore-minimum-tests/site-packages/numba/core/config.py:20: in <module>
    MACHINE_BITS = tuple.__itemsize__ * 8
E   AttributeError: type object 'tuple' has no attribute '__itemsize__'
=============================== warnings summary ==========================
```

I may move the doc yaml file to 3.* or 3.8 as it's currently as 3.6

> maybe you could also add a test for the epsilon case

Maybe not be a bad idea. I Could test with a DataArray with some 0's sprinkled in. What would be the best way to do this? create an `o_with_zeros` in the conftest?

aaronspring commented:
> Maybe not be a bad idea. I Could test with a DataArray with some 0's sprinkled in. What would be the best way to do this? create an o_with_zeros in the conftest?

yes

raybellwaves commented:
@aaronspring  What's the syntax for parametrizing a variable from the conftest?

e.g. trying

```
@pytest.mark.parametrize("a_1d", [a_1d, a_1d_with_zeros])
def test_xs_same_as_skl_mape(a_1d, b_1d):
    actual = mape(a_1d, b_1d, "time")
    expected = mean_absolute_percentage_error(a_1d, b_1d)
    assert np.allclose(actual, expected)
```

but says a_1d is not defined. It picks up a_1d in
```
def test_xs_same_as_skl_mape(a_1d, b_1d):
    actual = mape(a_1d, b_1d, "time")
    expected = mean_absolute_percentage_error(a_1d, b_1d)
    assert np.allclose(actual, expected)
```

I guess in the parametrize line it doesn't know `a_1d` and `a_1d_with_zeros` come from the conftest

raybellwaves commented:
I also changed some other things to possible help with the CI. Where we had a GHA testing again 3.6 I moved it to 3.8. I also updated the doc yaml to 3.* from 3.6.

raybellwaves commented:
I rewrote the test so I think i'm good there

aaronspring commented:
Probably fix this numba thing in a separate PR as it persists with 3.6.

maybe just taking a more recent version works, although I didn’t find the cause in https://github.com/numba/numba/blob/master/CHANGE_LOG

aaronspring commented on ci/minimum-tests.yml:
```diff
@@ -5,7 +5,7 @@ dependencies:
   - cftime
   - dask
-  - numba
+  - numba==0.51.2
```
```suggestion
  - numba
```

aaronspring commented:
`MACHINE_BITS = tuple.__itemsize__ * 8`
https://github.com/numba/numba/blob/ba0a6c8e7a7e33ad812bc644b29d2dfd8f3f4f0d/numba/core/config.py#L20

cannot reproduce locally. `tuple` is from python, so there seems to be some problem in GHA?

raybellwaves commented:
> cannot reproduce locally. tuple is from python, so there seems to be some problem in GHA?

Yeah it's probably the architecture GHA runs on e.g. linux. I believe there is a place where you can report it for the GHA team to look at it.

We can either merge this knowing that the py3.6 CI is bogus or remove py3.6 from the test suite.

raybellwaves commented:
@aaronspring if you don't mind, for the future can you do your test commits on a branch off of mine rather than this PR

aaronspring commented:
what do we do with that failing py36 testing?

aaronspring commented on xskillscore/tests/test_metric_results_accurate.py:
```diff
@@ -31,6 +31,10 @@
+    (mape, mean_absolute_percentage_error),
+]
+
 xs_scipy_metrics = [
```
thats not needed anymore, right?

raybellwaves replied:
yeah it used in the loop over metrics we want to test again when they encounter 0. MAPE is the only relevant one at the moment but as you mentioned we could add in the epsilon in other metrics in the future and expand this list.

aaronspring commented on xskillscore/tests/test_metric_results_accurate.py:
```diff
@@ -74,12 +78,10 @@ def test_xs_same_as_skl_same_name(a_1d, b_1d, request):
-        getattr(sklearn.metrics, request.param),
-    )
+@pytest.mark.parametrize("xs_skl_metrics", xs_skl_metrics_with_zeros)
+def test_xs_same_as_skl_with_zeros(a_1d_with_zeros, b_1d, xs_skl_metrics):
```
nice, this should have failed for mape before

raybellwaves replied:
yeah I think so

raybellwaves commented:
> what do we do with that failing py36 testing?

If you are happy to we can merge this.

I'll create a separate issue for the py3.6 CI failing.

aaronspring commented:
Just remove my small try in the squash merge

raybellwaves commented:
> Just remove my small try in the squash merge

I think it's ok. I don't really care about the commit history. As long as the files changed are desired.
